vfio_user/src/lib.rs

```diff
@@ -0,0 +1,675 @@
+        Self {
+            max_msg_fds: default_max_msg_fds(),
+            max_data_xfer_size: default_max_data_xfer_size(),
+            migration: default_migration_capabilities(),
+        }
+    }
+}
+
+pub struct Client {
+    stream: UnixStream,
+    next_message_id: Wrapping<u16>,
+    num_irqs: u32,
+    resettable: bool,
+    regions: Vec<Region>,
+}
+
+#[derive(Debug)]
+pub struct Region {
+    pub flags: u32,
+    pub index: u32,
+    pub size: u64,
+    pub file_offset: Option<FileOffset>,
+}
+
+#[derive(Debug)]
+pub struct IrqInfo {
+    pub index: u32,
+    pub flags: u32,
+    pub count: u32,
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Error connecting: {0}")]
+    Connect(#[source] std::io::Error),
+    #[error("Error serializing capabilities: {0}")]
+    SerializeCapabilites(#[source] serde_json::Error),
+    #[error("Error deserializing capabilities: {0}")]
+    DeserializeCapabilites(#[source] serde_json::Error),
+    #[error("Error writing to stream: {0}")]
+    StreamWrite(#[source] std::io::Error),
+    #[error("Error reading from stream: {0}")]
+    StreamRead(#[source] std::io::Error),
+    #[error("Error writing with file descriptors: {0}")]
+    SendWithFd(#[source] vmm_sys_util::errno::Error),
+    #[error("Error reading with file descriptors: {0}")]
+    ReceiveWithFd(#[source] vmm_sys_util::errno::Error),
+    #[error("Not a PCI device")]
+    NotPciDevice,
+}
+
+impl Client {
+    pub fn new(path: &Path) -> Result<Client, Error> {
+        let stream = UnixStream::connect(path).map_err(Error::Connect)?;
+
+        let mut client = Client {
+            next_message_id: Wrapping(0),
+            stream,
+            num_irqs: 0,
+            resettable: false,
+            regions: Vec::new(),
+        };
+
+        client.negotiate_version()?;
+
+        client.regions = client.get_regions()?;
+
+        Ok(client)
+    }
+
+    fn negotiate_version(&mut self) -> Result<(), Error> {
+        let caps = Capabilities::default();
+
+        let version_data = serde_json::to_string(&caps).map_err(Error::SerializeCapabilites)?;
+
+        let version = Version {
+            header: Header {
+                message_id: self.next_message_id.0,
+                command: Command::Version,
+                flags: HeaderFlags::Command as u32,
+                message_size: (std::mem::size_of::<Version>() + version_data.len() + 1) as u32,
+                ..Default::default()
+            },
+            major: 0,
+            minor: 1,
+        };
+        info!("Command: {:?}", version);
+
+        let version_data = CString::new(version_data.as_bytes()).unwrap();
+        let bufs = vec![
+            IoSlice::new(version.as_slice()),
+            IoSlice::new(version_data.as_bytes_with_nul()),
+        ];
+
+        // TODO: Use write_all_vectored() when ready
+        let _ = self
+            .stream
+            .write_vectored(&bufs)
+            .map_err(Error::StreamWrite)?;
```

Comment:
So you're not trying to handle the case where the write would return less than expected?

Reply:
Not at the moment. For small writes to a UNIX domain socket the writes will be buffered so will complete unless there is a peer disconnect at which point the error will trigger. For completeness I would like to use `write_all_vectored` when it becomes available but it is not worth the complexity of dealing with partial writes across vectored writes. Using vector write is essential since otherwise the libvfio-use code will think it only gets a partial read and reject the command.

Reply:
Sounds good!